### PR TITLE
[Doctests] Move doctests to new GPU & Fix bugs

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This fixes all the doc tests and moves them to a CPU runner. 
99% of doc-tests are run on CPU anyways in PyTorch and it shouldn't make a difference in Tensorflow whether they are on CPU or GPU, so I think to save some $2000 per month we can safely run them on CPU.

Ok for you @sgugger @LysandreJik ?

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: .github/workflows/doctests.yml [hunk 1]
```diff
@@ -5,8 +5,16 @@ on:
     branches:
       - doctest*
   repository_dispatch:
-  schedule:
-    - cron: "0 0 * * *"
+  pull_request:
```
Comment (patrickvonplaten) [inline .github/workflows/doctests.yml:8]: Will revert this before merge

File: .github/workflows/doctests.yml [hunk 2]
```diff
@@ -19,32 +27,30 @@ env:
 
 jobs:
   run_doctests:
-    runs-on: [self-hosted, docker-gpu-test, single-gpu]
+    runs-on: [self-hosted, doc-tests]
```
Comment (patrickvonplaten) [inline .github/workflows/doctests.yml:30]: created a new CPU runner

File: docs/source/quicktour.mdx [hunk 3]
```diff
@@ -99,12 +99,16 @@ The [`pipeline`] can also iterate over an entire dataset. Start by installing th
...
-Create a [`pipeline`] with the task you want to solve for and the model you want to use. Set the `device` parameter to `0` to place the tensors on a CUDA device:
+Create a [`pipeline`] with the task you want to solve for and the model you want to use. Set the `device` parameter to `0` to place the tensors on a CUDA device if you
+it is available:
 
 ```py
+>>> import torch
 >>> from transformers import pipeline
 
->>> speech_recognizer = pipeline("automatic-speech-recognition", model="facebook/wav2vec2-base-960h", device=0)
+>>> device = 0 if torch.cuda.is_available() else -1
```
Comment (patrickvonplaten) [inline docs/source/quicktour.mdx:109]: This is the only time where we explicitly say in the doc tests that we should run on GPU. I think it's fine here to add an additional `if torch.cuda.is_availabsle()` to make it also pass on CPU. Ok for you @sgugger ?

File: src/transformers/models/beit/modeling_beit.py [hunk 4]
```diff
@@ -55,7 +55,7 @@
 
 # Image classification docstring
 _IMAGE_CLASS_CHECKPOINT = "microsoft/beit-base-patch16-224"
-_IMAGE_CLASS_EXPECTED_OUTPUT = "'tabby, tabby cat'"
+_IMAGE_CLASS_EXPECTED_OUTPUT = "tabby, tabby cat"
```
Comment (patrickvonplaten) [inline src/transformers/models/beit/modeling_beit.py:58]: @NielsRogge the test failed

File: src/transformers/models/deit/modeling_deit.py [hunk 5]
```diff
@@ -697,9 +697,11 @@ def forward(
 
         ```python
         >>> from transformers import DeiTFeatureExtractor, DeiTForImageClassification
+        >>> import torch
         >>> from PIL import Image
         >>> import requests
 
+        >>> torch.manual_seed(3)  # doctest: +IGNORE_RESULT
```
Comment (patrickvonplaten) [inline src/transformers/models/deit/modeling_deit.py:704]: random weights are loaded - we need a `manual_seed(3)` here

File: src/transformers/models/poolformer/modeling_poolformer.py [hunk 6]
```diff
@@ -44,11 +44,11 @@
 
 # Base docstring
 _CHECKPOINT_FOR_DOC = "sail/poolformer_s12"
-_EXPECTED_OUTPUT_SHAPE = [1, 197, 768]
+_EXPECTED_OUTPUT_SHAPE = [1, 512, 7, 7]
```
Comment (patrickvonplaten) [inline src/transformers/models/poolformer/modeling_poolformer.py:47]: @NielsRogge - this was failing as well. Is the new output shape ok?

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 7]
```diff
@@ -1611,7 +1611,6 @@ def __init__(self, config):
         self.post_init()
 
     @add_start_docstrings_to_model_forward(WAV_2_VEC_2_INPUTS_DOCSTRING)
-    @replace_return_docstrings(output_type=Wav2Vec2BaseModelOutput, config_class=_CONFIG_FOR_DOC)
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:1614]: This class is deprecated and we don't document it anymore. So removing the docstring here completely as it was failing

**Comment by patrickvonplaten** [comment]
@sgugger @LysandreJik - if you're fine with new CPU doc test runner, then if possible it would also be nice to report test failures just like we do the daily slow tests. 

Would be happy to help set it up

File: src/transformers/models/poolformer/modeling_poolformer.py (hunk 6 above)
Comment (NielsRogge) [reply src/transformers/models/poolformer/modeling_poolformer.py:47]: Yes that looks ok, thanks for fixing.

File: src/transformers/models/beit/modeling_beit.py (hunk 4 above)
Comment (NielsRogge) [reply src/transformers/models/beit/modeling_beit.py:58]: Thanks for fixing

File: docs/source/quicktour.mdx [hunk 8]
```diff
@@ -99,12 +99,16 @@ The [`pipeline`] can also iterate over an entire dataset. Start by installing th
 pip install datasets 
 ```
 
-Create a [`pipeline`] with the task you want to solve for and the model you want to use. Set the `device` parameter to `0` to place the tensors on a CUDA device:
+Create a [`pipeline`] with the task you want to solve for and the model you want to use. Set the `device` parameter to `0` to place the tensors on a CUDA device if you
+it is available:
```
Comment (sgugger) [inline docs/source/quicktour.mdx:103]:
```suggestion
Create a [`pipeline`] with the task you want to solve for and the model you want to use. Use the `device` parameter to place the tensors on a GPU if you have one:
```

File: docs/source/quicktour.mdx (hunk 3 above)
Comment (sgugger) [reply docs/source/quicktour.mdx:109]:
```suggestion
>>> device = "cuda" if torch.cuda.is_available() else "cpu"
```
This is clearer.

File: docs/source/quicktour.mdx [hunk 9]
```diff
@@ -264,10 +268,10 @@ tensor([[0.0021, 0.0018, 0.0115, 0.2121, 0.7725],
 >>> import tensorflow as tf
 
 >>> tf_predictions = tf.nn.softmax(tf_outputs.logits, axis=-1)
->>> print(tf_predictions)
+>>> print(tf.math.round(tf_predictions * 10**4) / 10**4)
```
[the remark this answers is not in the document]
Comment (sgugger) [reply docs/source/quicktour.mdx:271]: No, doctest will round all numbes to the precision written, whether inside tensors or not. This passed locally before so should still pass by tweaking the display. We should avoid cluttering the example for the sake of the doctests and can just ignore the output of this particular test if needs be.

**Review by sgugger — APPROVED** [review]
Thanks for fixing! I'm just not for making the examples more complex (especially in the quicktour) just for the sake of the doctests. Apart from that, it all looks good to me!

Note that moving the job to a CPU means we won't test any of the training tutorials however, so not sure it's worth it. But we can switch back when those pass.

File: docs/source/quicktour.mdx (hunk 9 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply docs/source/quicktour.mdx:271]: Didn't pass for me locally - could you maybe quickly double check here? Note that all tests were passing locally before because we were ignoring the output.

If they pass locally for you, happy to revert it. PT passes because they automatically round to 4 decimal places

File: docs/source/quicktour.mdx (hunk 3 above)
Comment (patrickvonplaten) [reply docs/source/quicktour.mdx:109]: Think this doesn't work no @Narsil ?

**Comment by patrickvonplaten** [comment]
> Thanks for fixing! I'm just not for making the examples more complex (especially in the quicktour) just for the sake of the doctests. Apart from that, it all looks good to me!
> 
> Note that moving the job to a CPU means we won't test any of the training tutorials however, so not sure it's worth it. But we can switch back when those pass.

Yeah, I see the point here! I was thinking a bit too much about just the model docstring examples which all always run on CPU. Happy to switch it back once we have the training examples working!

File: docs/source/quicktour.mdx (hunk 3 above)
Comment (sgugger) [reply docs/source/quicktour.mdx:109]: Indeed. Let's remove GPU entirely then as it's not natural at all for a newcomer from PyTorch (@Narsil why doesn't `device` support the same thing as `torch.device`? This is very confusing...)

Comment (patrickvonplaten) [reply docs/source/quicktour.mdx:109]: But agree with @sgugger here that it should work. It's weird that `device` has to be an `int` here

Comment (Narsil) [reply docs/source/quicktour.mdx:109]: How, that is old code being moved over and over I think.
I guess initially it was done that way to be roughly agnostic to TF vs PT.

I do agree that being at least compatible with native things is OK (it should be at least).

For the example, I guess it depends where we are, but it's nice when the pipeline uses whatever backend is installed on your end (so not relying on `torch.device` nor `tf.device('/device:GPU:2')` by default but a simple subset for simple usage)

Comment (sgugger) [reply docs/source/quicktour.mdx:109]: Oh yes, I hadn't even realized we are putting some framework-specific code in the generic example. All the more reason to remove the `device` altogether, we shouldn't import `torch` here (or have some TF block that does the equivalent, but for this is starting to get too complex for the first part of the quicktour).

Comment (Narsil) [reply docs/source/quicktour.mdx:109]: Let's remove the `device` altogether ! Fine by me.

File: docs/source/quicktour.mdx [hunk 10]
```diff
@@ -264,6 +265,8 @@ tensor([[0.0021, 0.0018, 0.0115, 0.2121, 0.7725],
 >>> import tensorflow as tf
 
 >>> tf_predictions = tf.nn.softmax(tf_outputs.logits, axis=-1)
+>>> # print(tf.math.round(tf_predictions * 10**4) / 10**4)
```
Comment (patrickvonplaten) [inline docs/source/quicktour.mdx:268]:
```suggestion
```

File: .github/workflows/doctests.yml [hunk 11]
```diff
@@ -16,35 +24,50 @@ env:
...
-    runs-on: [self-hosted, docker-gpu-test, single-gpu]
+    runs-on: [self-hosted, doc-tests-gpu]
     container:
-      image: pytorch/pytorch:1.9.0-cuda11.1-cudnn8-runtime
+      image: huggingface/transformers-all-latest-gpu
       options: --gpus 0 --shm-size "16gb" --ipc host -v /mnt/cache/.cache/huggingface:/mnt/cache/
     steps:
-      - name: Launcher docker
-        uses: actions/checkout@v2
+      - name: Remove transformers
```
Comment (patrickvonplaten) [inline .github/workflows/doctests.yml:37]: is this correct here @LysandreJik ?

**Comment by patrickvonplaten** [comment]
Think everything works as expected now! Tests will run daily on GPU. Thanks for the help @LysandreJik

File: docs/source/quicktour.mdx (hunk 9 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply docs/source/quicktour.mdx:268]: Happy to fix in a follow-up PR! Didn't manage to make it work without rounding (neither locally nor on the ci-runner, see message offline).

**Comment by patrickvonplaten** [comment]
Cleaning up the rounding stuff in the docs in a follow-up PR
